src/common/src/util/chunk_coalesce.rs

```diff
@@ -157,17 +157,27 @@ impl DataChunkBuilder {
     /// Return a data chunk if the buffer is full after append one row. Otherwise `None`.
     #[must_use]
     pub fn append_one_row(&mut self, row: impl Row) -> Option<DataChunk> {
-        assert!(self.buffered_count < self.batch_size);
-        self.ensure_builders();
-
-        self.do_append_one_row_from_datums(row.iter());
+        self.append_one_row_no_finish(row);
         if self.buffered_count == self.batch_size {
             Some(self.build_data_chunk())
         } else {
             None
         }
     }
 
+    /// Append one row from the given [`Row`].
+    /// The caller is responsible to build the data chunk,
```

Comment:
I think this somehow breaks the original design of `DataChunkBuilder`. Actually, developers may call `assert!(append_one_row(..).is_none())` to achieve the same behavior. 🤔

Reply:
`append_one_row()` also calls `build_data_chunk`:
```rs
    pub fn append_one_row(&mut self, row: impl Row) -> Option<DataChunk> {
        self.append_one_row_no_finish(row);
        if self.buffered_count == self.batch_size {
            Some(self.build_data_chunk())
        } else {
            None
        }
    }
```

We are calling this function repeatedly when iterating over chunk rows, so I prefer to just call `build_data_chunk` once at the end.

Reply:
Hmm I guess in terms of safety this approach will be better. I will fix it and leave a TODO if we might need to optimize it in the future.